### PR TITLE
feat: Added an alert cache mechanism to avoid loosing information without internet

### PR DESCRIPTION
**Pull request by frgfm**
This PR introduces the following modifications:
- unified alert uploading: any detected alert is put in a cache (with a max size), and the object tries to empty this cache by uploading the alerts in it at every prediction it does. This means that if the connection is lost during N detections and comes back after this, it will dutifully upload the media and send the alerts for those in cache (limited to the cache size obviously)
- added an automatic cache backup system: this cache is backup to disk every hour by default. Everytime the engine is instantiated it tries to restore the cache from disk. Which means that when we reboot the device to get back the connection, it will load back everything that was saved to disk.
- updates requirements
- updates unittests

One point to pay attention at: this can be checked in another PR but I prefer to point it out now. This PR handles the process as if the disk is persistent through reboot (if I save a file in `my_path.jpg`, then reboot, I'll find it in the same place). However, since we are using Docker and I'm not too sure about whether we bind volumes in production, this might be tricky. Anyone got more information about this? (I set the back path to `data/` which seems to be the volume bind according to https://github.com/pyronear/pyro-engine/blob/master/server/docker-compose.yml#L14)

Any feedback is welcome!

### CODE REVIEW

**Review by blenzi — APPROVED** [review]
Hi @frgfm , thanks a lot for this PR. It looks good to me and I don't have more information about the binding of the volumes. I guess it all needs to be checked manually, right?

**Review by Akilditu — COMMENTED** [review]
Hi thanks a lot for this PR ! 

If I understand correctly, your concern about the reboot is the ensure that the `data` volume remains persistent after a reboot of the main rpi which rebuilds docker ? 
A dump of cache outside of docker could then be a solution ? 

Another question here, could we estimate on avg how many 'lost connection time' we can recover having 100 frames within a full alert time-lapse ?

File: pyroengine/engine/engine.py [hunk 1]
```diff
@@ -16,130 +25,214 @@ class PyronearEngine:
...
-        longitude=None
-    ):
+        detection_thresh: float = 0.5,
+        api_url: Optional[str] = None,
+        client_creds: Optional[Dict[str, str]] = None,
+        frame_saving_period: Optional[int] = None,
+        latitude: Optional[float] = None,
+        longitude: Optional[float] = None,
+        cache_size: int = 100,
+        alert_relaxation: int = 3,
```
[the remark this answers is not in the document]
Comment (frgfm) [reply pyroengine/engine/engine.py:54]: Correct!

File: pyroengine/engine/engine.py [hunk 2]
```diff
@@ -16,130 +25,214 @@ class PyronearEngine:
...
-        save_every_n_frame (int): Send one frame over N to the api for our dataset
-        latitude (float): device latitude
-        longitude (float): device longitude
+        detection_thresh: wildfire detection threshold in [0, 1]
+        api_url: url of the pyronear API
+        client_creds: api credectials for each pizero, the dictionary should as the one in the example
+        frame_saving_period: Send one frame over N to the api for our dataset
+        latitude: device latitude
+        longitude: device longitude
+        cache_size: maximum number of alerts to save in cache
```
[the remark this answers is not in the document]
Comment (frgfm) [reply pyroengine/engine/engine.py:34]: Yes, but those frames are only those that raised the alerts

**Comment by frgfm** [comment]
> If I understand correctly, your concern about the reboot is the ensure that the `data` volume remains persistent after a reboot of the main rpi which rebuilds docker ?

Correct

> A dump of cache outside of docker could then be a solution ?

Yes, that's why I put this in `data/` which is the default volume location

> Another question here, could we estimate on avg how many 'lost connection time' we can recover having the 100 frames max storage in cache within a full alert time-lapse ?

Rather easy yes :
each raspberry can at maximum detect fire each 30sec, say we have N of these:
which makes 2 * N frames per minute
So with 100 frames, the lower bound (in case of continuous wildfire detection at every analysis) we could cover is 100 / (2 * N) minutes. So with 4 active cameras, this covers 12 minutes

But this would be the very worst case scenario (all devices won't be pointing towards the same location, and hopefully we won't have lost connection for all of those) :)

**Comment by Akilditu** [comment]
> Yes, that's why I put this in `data/` which is the default volume location

Isn't '/data' gonna be overridden if the docker re-builds ?

**Review by fe51 — COMMENTED** [review]
So far, it looks good to me. I would like to set up test devices to approve, if I can find the time.

About volume binding, as far as I understand, the volume source is [./data](https://github.com/pyronear/pyro-engine/blob/master/server/docker-compose.yml#L14) so on the host, hence implicitly of type "bind".

**Comment by frgfm** [comment]
> > Yes, that's why I put this in `data/` which is the default volume location
> 
> Isn't `data/` gonna be overridden if the docker re-builds ?

So I've checked on my own, having a docker with a volume bind, creating a file in it, and then rebuild the image and check whether the file is still there and it works!

With this docker-compose:
```
version: '3.7'

services:
  web:
    image: python:3.8.1-slim
    command: >
        bash -c "ls /tmp
        && touch /tmp/tmp.txt
        && python -c 'print(1)'
        && ls /tmp
        "
    ports:
      - 8050:8000
    volumes:
      - /tmp:/home/fg/Downloads/tmp

```

I run it the first time to create the file in the volume:
```shell
➜  docker_check docker-compose up --build

Recreating docker_check_web_1 ... done
Attaching to docker_check_web_1
web_1  | 1
web_1  | tmp.txt
docker_check_web_1 exited with code 0
```

then a second time to see if it can access the created file:
```
➜  docker_check docker-compose up --build

Starting docker_check_web_1 ... done
Attaching to docker_check_web_1
web_1  | tmp.txt
web_1  | 1
web_1  | tmp.txt
docker_check_web_1 exited with code 0
```

**Review by fe51 — APPROVED** [review]
Thanks for the check FG ! So, let's merge 😃

**Pull request merged**